SimPEG/PF/Magnetics.py

```diff
@@ -23,361 +26,476 @@ class MagneticIntegral(Problem.LinearProblem):
+        # (lower and upper coners for each cell)
+        if isinstance(self.mesh, Mesh.TreeMesh):
+            # Get upper and lower corners of each cell
+            bsw = (self.mesh.gridCC - self.mesh.h_gridded/2.)
+            tne = (self.mesh.gridCC + self.mesh.h_gridded/2.)
 
-            return fwr_d
+            xn1, xn2 = bsw[:, 0], tne[:, 0]
+            yn1, yn2 = bsw[:, 1], tne[:, 1]
+            zn1, zn2 = bsw[:, 2], tne[:, 2]
 
         else:
 
-            # m = np.hstack([m, mii])
+            xn = self.mesh.vectorNx
+            yn = self.mesh.vectorNy
+            zn = self.mesh.vectorNz
 
-            return self.G.dot(m)
+            yn2, xn2, zn2 = np.meshgrid(yn[1:], xn[1:], zn[1:])
+            yn1, xn1, zn1 = np.meshgrid(yn[:-1], xn[:-1], zn[:-1])
 
-    @property
-    def G(self):
-        if not self.ispaired:
-            raise Exception('Need to pair!')
+        # If equivalent source, use semi-infite prism
+        if self.equiSourceLayer:
+            zn1 -= 1000.
 
-        if getattr(self, '_G', None) is None:
-            self._G = self.Intrgl_Fwr_Op(Magnetization='xyz')
+        self.Yn = P.T*np.c_[mkvc(yn1), mkvc(yn2)]
+        self.Xn = P.T*np.c_[mkvc(xn1), mkvc(xn2)]
+        self.Zn = P.T*np.c_[mkvc(zn1), mkvc(zn2)]
 
-        return self._G
+        # survey = self.survey
+        self.rxLoc = self.survey.srcField.rxList[0].locs
 
+        if magType == 'H0':
+            if getattr(self, 'M', None) is None:
+                self.M = matutils.dipazm_2_xyz(np.ones(nC) * self.survey.srcField.param[1],
+                                      np.ones(nC) * self.survey.srcField.param[2])
 
-class MagneticAmplitude(MagneticIntegral):
+            Mx = sdiag(self.M[:, 0] * self.survey.srcField.param[0])
+            My = sdiag(self.M[:, 1] * self.survey.srcField.param[0])
+            Mz = sdiag(self.M[:, 2] * self.survey.srcField.param[0])
 
-    forwardOnly = False  # If false, matric is store to memory (watch your RAM)
-    actInd = None  #: Active cell indices provided
-    M = None  #: Magnetization matrix provided, otherwise all induced
-    rtype = 'xyz'  #: Receivers must be "xyz"
+            self.Mxyz = sp.vstack((Mx, My, Mz))
 
-    def __init__(self, mesh, **kwargs):
-        Problem.BaseProblem.__init__(self, mesh, **kwargs)
+        elif magType == 'full':
 
-    def fwr_ind(self, m):
+            self.Mxyz = sp.identity(3*nC) * self.survey.srcField.param[0]
 
-        self.survey.srcField.rxList[0].rxType = 'xyz'
+        else:
+            raise Exception('magType must be: "H0" or "full"')
 
-        if self.forwardOnly:
+                # Loop through all observations and create forward operator (nD-by-nC)
+        print("Begin forward: M=" + magType + ", Rx type= " + self.rxType)
 
-            # Compute the linear operation without forming the full dense G
-            Bxyz = Intrgl_Fwr_Op(m=m)
+        # Switch to determine if the process has to be run in parallel
+        job = Forward(
```

Comment:
sorry to be picky, would you mind closing the bracket at the same level as the `job` e.g.
```
job = Forward(
    rxLoc= ... 
)
```